clash-prelude/tests/Clash/Tests/Laws/Enum.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE RankNTypes #-}
+
+module Clash.Tests.Laws.Enum (tests) where
+
+import Control.DeepSeq (NFData)
+import Data.Proxy
+import Test.Tasty
+import Test.Tasty.HUnit
+
+import Clash.Sized.Index (Index)
+import Clash.Sized.Signed (Signed)
+import Clash.Sized.Fixed (SFixed, UFixed)
+import Clash.Sized.Unsigned (Unsigned)
+
+import Test.Tasty.HUnit.Extra
+
+succMaxBoundLaw ::
+  forall a .
+  (NFData a, Show a, Enum a, Bounded a) =>
+  Proxy a ->
+  Assertion
+succMaxBoundLaw Proxy = expectException (succ @a maxBound)
+
+predMinBoundLaw ::
+  forall a .
+  (NFData a, Show a, Enum a, Bounded a) =>
+  Proxy a ->
+  Assertion
+predMinBoundLaw Proxy = expectException (pred @a minBound)
+
+enumLaws ::
+  (NFData a, Show a, Enum a, Bounded a) =>
+  Proxy a ->
+  [TestTree]
+enumLaws proxy =
+  [ testCase "succ maxBound ~ _|_" (succMaxBoundLaw proxy)
+  , testCase "pred minBound ~ _|_" (predMinBoundLaw proxy)
+  ]
+
+testEnumLaws ::
+  (NFData a, Show a, Enum a, Bounded a) =>
+  String ->
+  Proxy a ->
+  TestTree
+testEnumLaws typeName proxy = testGroup typeName (enumLaws proxy)
+
+tests :: TestTree
+tests = testGroup "Enum"
+  [ testEnumLaws "Index 1" (Proxy @(Index 1))
+  , testEnumLaws "Index 2" (Proxy @(Index 2))
+  , testEnumLaws "Index 128" (Proxy @(Index 128))
+
+  , testEnumLaws "Unsigned 0" (Proxy @(Unsigned 0))
+  , testEnumLaws "Unsigned 1" (Proxy @(Unsigned 1))
+  , testEnumLaws "Unsigned 32" (Proxy @(Unsigned 32))
+  , testEnumLaws "Unsigned 127" (Proxy @(Unsigned 127))
+  , testEnumLaws "Unsigned 128" (Proxy @(Unsigned 128))
+
+  , testEnumLaws "Signed 0" (Proxy @(Signed 0))
+  , testEnumLaws "Signed 1" (Proxy @(Signed 1))
+  , testEnumLaws "Signed 32" (Proxy @(Signed 32))
+  , testEnumLaws "Signed 127" (Proxy @(Signed 127))
+  , testEnumLaws "Signed 128" (Proxy @(Signed 128))
+
+  , testEnumLaws "SFixed 0 0" (Proxy @(SFixed 0 0))
+  , testEnumLaws "SFixed 0 1" (Proxy @(SFixed 0 1))
+  , testEnumLaws "SFixed 1 0" (Proxy @(SFixed 1 0))
+  , testEnumLaws "SFixed 1 1" (Proxy @(SFixed 1 1))
+  , testEnumLaws "SFixed 1 2" (Proxy @(SFixed 1 2))
+  , testEnumLaws "SFixed 2 1" (Proxy @(SFixed 2 1))
+  , testEnumLaws "SFixed 2 2" (Proxy @(SFixed 2 2))
+  , testEnumLaws "SFixed 128 128" (Proxy @(SFixed 128 128))
+
+  , testEnumLaws "UFixed 0 0" (Proxy @(UFixed 0 0))
+  , testEnumLaws "UFixed 0 1" (Proxy @(UFixed 0 1))
+  , testEnumLaws "UFixed 1 0" (Proxy @(UFixed 1 0))
+  , testEnumLaws "UFixed 1 1" (Proxy @(UFixed 1 1))
+  , testEnumLaws "UFixed 1 2" (Proxy @(UFixed 1 2))
+  , testEnumLaws "UFixed 2 1" (Proxy @(UFixed 2 1))
+  , testEnumLaws "UFixed 2 2" (Proxy @(UFixed 2 2))
+  , testEnumLaws "UFixed 128 128" (Proxy @(UFixed 128 128))
```

Comment:
Hmm.. this is a weird one. The law says:

> The calls succ maxBound and pred minBound should result in a runtime error.

I.e., it says nothing about values other than `maxBound` and `minBound`. So strictly, I think the tests are "exactly right". But you're point stands: maybe we should just make more tests specific for `SFixed` / `UFixed`?

Reply:
Well, I think the law was written without considering anybody would ever add a bounded fractional type :-). Because if `succ` means add 1, you have a type that ranges up to 15.75, and you do `succ 15.5` what else is it going to do than runtime error? You're trying to shove the value 16.5 into a type that can't represent that value. So I think we need to interpret the law as parliament intended, not as what they literally wrote.

Reply:
I've disabled those specific tests for Fixed, as your testsuite covers it :).